### PR TITLE
Trigger checkpoints and restoring in smoke tests

### PR DESCRIPTION
**Pull request by jbeisen**
Modify the single file source to block on receiving control messages, which enables the smoke tests to deterministically trigger checkpoints, and use a Noop control message to advance the source.

Running successive engines in the same process required changing how Prometheus metric registration works because metrics cannot be registered multiple times. This changes the worker to statically register counter vectors and apply labels when the counter is used.

This also splits up the job controller state into separate modules.

A later commit will add state compaction to the smoke tests.

### CODE REVIEW

File: arroyo-controller/src/job_controller/checkpoint_state.rs
Comment (jbeisen) [inline arroyo-controller/src/job_controller/checkpoint_state.rs:1]: This is all just moved from `job_controller/checkpointer.rs`.

File: arroyo-sql-testing/src/smoke_tests.rs [hunk 1]
```diff
@@ -21,14 +121,56 @@ async fn run_pipeline_and_assert_outputs(
...
+    let engine = Engine::for_local(program, job_id.clone());
+    let (running_engine, mut control_rx) = engine
         .start(StreamConfig {
             restore_epoch: None,
         })
         .await;
-    // wait for the engine to finish
-    while let Some(control_resp) = control_rx.recv().await {}
+
+    if checkpoints {
```
Comment (jacksonrnewhouse) [inline arroyo-sql-testing/src/smoke_tests.rs:133]: Can this be factored out into a method?

File: arroyo-sql-macro/src/lib.rs [hunk 2]
```diff
@@ -209,6 +209,7 @@ impl Parse for PipelineCase {
 
 struct CorrectnessTestCase {
     test_name: LitStr,
+    checkpoints: LitStr,
```
Comment (jacksonrnewhouse) [inline arroyo-sql-macro/src/lib.rs:212]: You can make this a `syn::LitBool` and then your macros can just take `true` and `false`.

File: arroyo-worker/src/engine.rs [hunk 3]
```diff
@@ -149,33 +148,30 @@ impl OutQueue {
...
             let bytes = bincode::encode_to_vec(&message, config::standard()).unwrap();
-            sent_bytes.iter().for_each(|c| c.inc_by(bytes.len() as u64));
+            TaskCounters::BytesSent
+                .for_task(task_info)
+                .inc_by(bytes.len() as u64);
 
             QueueItem::Bytes(bytes)
         } else {
-            QueueItem::Data(Box::new(message))
+            QueueItem::Data(Box::new(message.clone()))
```
Comment (jacksonrnewhouse) [inline arroyo-worker/src/engine.rs:161]: This clone should be avoided.

File: arroyo-sql-testing/src/smoke_tests.rs (hunk 1 above)
Comment (jbeisen) [reply arroyo-sql-testing/src/smoke_tests.rs:179]: Done

File: arroyo-sql-macro/src/lib.rs (hunk 2 above)
Comment (jbeisen) [reply arroyo-sql-macro/src/lib.rs]: Done

File: arroyo-worker/src/connectors/websocket.rs [hunk 4]
```diff
@@ -115,6 +115,7 @@ where
             ControlMessage::LoadCompacted { compacted } => {
                 ctx.load_compacted(compacted).await;
             }
+            ControlMessage::Noop => {}
```
Comment (jacksonrnewhouse) [inline arroyo-worker/src/connectors/websocket.rs:118]: I think this would be a NoOp typically

Comment (jbeisen) [reply arroyo-worker/src/connectors/websocket.rs]: Changed

File: arroyo-controller/src/job_controller/checkpoint_state.rs [hunk 5]
```diff
@@ -0,0 +1,396 @@
...
+            start_time: to_micros(self.start_time),
+            finish_time: to_micros(finish_time),
+            min_epoch: self.min_epoch,
+            operator_ids: self.completed_operators.iter().cloned().collect(),
+        })
+        .await;
+
+        let operator_state = serde_json::to_value(&self.operator_details).unwrap();
+
+        if let Some(pool) = pool {
```
Comment (jacksonrnewhouse) [inline arroyo-controller/src/job_controller/checkpoint_state.rs:355]: This seems a bit of an anti-pattern. This is now two methods "finish saving operator state" and "update DB with new checkpoint status" and depending on how it is called you only do the second sometimes. Can you try factoring it out into two methods?

File: arroyo-sql-testing/Cargo.toml [hunk 6]
```diff
@@ -27,3 +28,7 @@ tokio = { version = "1.16", features = ["full"] }
 arrow = {workspace = true }
 arrow-array = {workspace = true }
 arrow-schema = {workspace = true }
+tracing = "0.1.37"
+test-log = { version = "0.2.12", default-features = false, features = ["trace"] }
+tracing-subscriber = {version = "0.3.17"}
+#env_logger = "0.10.0"
```
Comment (jacksonrnewhouse) [inline arroyo-sql-testing/Cargo.toml:34]: delete

File: arroyo-controller/src/job_controller/checkpoint_state.rs [hunk 7]
```diff
@@ -0,0 +1,396 @@
...
+                .any(|table| table.write_behavior() == TableWriteBehavior::CommitWrites)
+        {
+            self.subtasks_to_commit
+                .insert((c.operator_id.clone(), metadata.subtask_index));
+        }
+
+        let subtasks = self
+            .tasks
+            .get_mut(&c.operator_id)
+            .unwrap_or_else(|| panic!("Received finish event without start for {}", c.operator_id));
```
[the remark this answers is not in the document]
Comment (jbeisen) [reply arroyo-controller/src/job_controller/checkpoint_state.rs]: Sure, done

File: arroyo-controller/src/job_controller/checkpoint_state.rs (hunk 5 above)
Comment (jbeisen) [reply arroyo-controller/src/job_controller/checkpoint_state.rs]: Good idea, I've split this into two functions and changed `finish_checkpoint_if_done` to use them.

**Review by jacksonrnewhouse — APPROVED** [review]

**Pull request merged**